File: java/leetcode_grind/Day502.java
package leetcode_grind;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Day502 {
    // https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/description/
    static class Solution1 {
        public long countSubarrays(int[] nums, int k) {
            int maxElement = Arrays.stream(nums).max().getAsInt();
            List<Integer> indexesOfMaxELements = new ArrayList<>();
            long ans = 0;

            for (int i = 0; i < nums.length; i++) {
                if (nums[i] == maxElement) {
                    indexesOfMaxELements.add(i);
                }
                int freq = indexesOfMaxELements.size();
                if (freq >= k) {
                    ans += indexesOfMaxELements.get(freq - k) + 1;
                }
            }

            return ans;
        }
    }
}

File: src/leetcode_grind/day_502.rs
// https://leetcode.com/problems/count-subarrays-where-max-element-appears-at-least-k-times/description/
pub fn count_subarrays(nums: Vec<i32>, k: i32) -> i64 {
    let mut l = 0;
    let mx = *nums.iter().max().unwrap();
    let mut curr = 0;
    let mut ans = 0;

    for r in 0..nums.len() {
        if nums[r] == mx {
            curr += 1;
        }

        while curr == k {
            if nums[l] == mx {
                curr -= 1;
            }
            l += 1;
        }

        ans += l as i64;
    }

    ans
}
